modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,298 @@
+                streams[i] = null;
+                streamOff++;
+            }
+            else {
+                // Prefetch head value.
+                s.next();
+            }
+        }
+
+        if (finished())
+            throw new NoSuchElementException("No next element. Please, be sure to invoke hasNext() before next().");
+
+        if (first) {
+            first = false;
+
+            Arrays.sort(streams, streamCmp);
+        } else
+            bubbleUp(streams, streamOff, streamCmp);
+
+        return streams[streamOff].get();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() throws IgniteCheckedException {
+        for (int i = streamOff; i < streams.length; i++) {
+            if (streams[i].hasNext())
+                return true;
+
+            // Nullify obsolete stream, move left bound.
+            streamsMap.remove(streams[i].nodeId);
+            streams[i] = null;
+            streamOff++;
+        }
+
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void loadPage() {
+        assert !Thread.holdsLock(queueLock());
+
+        List<UUID> nodes;
+
+        synchronized (queueLock()) {
+            if (!loadAllowed)
+                return;
+
+            nodes = new ArrayList<>();
+
+            for (int i = streamOff; i < streams.length; i++) {
+                UUID nodeId = streams[i].nodeId;
+
+                // Try to contain 2 pages for every stream to avoid waits. A node has to be present in collections:
+                // 1. rcvd - as requested and already handled before, so currently no parallel request to this node.
+                // 2. subgrid - as this node still has pages to request.
+                if (streams[i].queue.size() < 1 && rcvd.remove(nodeId) && subgrid.contains(nodeId))
+                    nodes.add(nodeId);
+            }
+        }
+
+        if (nodes != null)
+            pageRequester.requestPages(reqId, fut, nodes, false);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onPage(@Nullable UUID nodeId, boolean last) {
+        boolean qryLast = super.onPage(nodeId, last);
+
+        if (nodeId == null)
+            nodeId = cctx.localNodeId();
+
+        if (last && streamsMap.containsKey(nodeId)) {
+            streamsMap.get(nodeId).allPagesReady = true;
+
+            streamsMap.remove(nodeId);
+        }
+
+        return qryLast;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addPage(@Nullable UUID nodeId, Collection<R> data) {
+        assert Thread.holdsLock(queueLock());
+
+        // Local node.
+        if (nodeId == null) {
+            nodeId = cctx.localNodeId();
+
+            // If nodeId is NULL and query doesn't execute on local node, then it is error, notify all streams.
+            if (!streamsMap.containsKey(nodeId)) {
+                assert data.isEmpty();
+
+                for (PageStream stream: streamsMap.values())
+                    stream.addPage(data);
```

Comment:
Why do you pass 'data' to all streams, if got a page from an unknown node?

Reply:
In this case we know that nodeId set to NULL only in case of error, that data is empty. So it's safe to put data. Note that we put this to all nodes to faster finish stream traverse to faster check future for error.

Reply:
So, why we have a magic constant NULL instead of a separate callback for this with error as a param?
It is totally unclear, why UnsortedReducer doesn't have a similar null check...

Also, passing an empty page looks like a magic constant as well.
Do we need to notify other streams? Do we know at this point, that prefetched result (failed) will ever be needed?